File: LeetCode/526-beautiful-arrangement/ex_526.cpp
class Solution {
public:
    int countArrangement(int N) {
        vector<int> v;
        for (auto i = 0; i != N; ++i)
            v.push_back(i+1);
        return backtrack(v, N);
    }
    
private: 
    int backtrack(vector<int> &v, int n) {
        if (n <= 0) return 1;
        int count = 0;
        for (auto i = 0; i < n; ++i) {
            if (v[i] % n == 0 or n % v[i] == 0) {
                swap(v[i], v[n-1]);
                count += backtrack(v, n-1);
                swap(v[i], v[n-1]);
            }
        }
        return count;
    }
};

File: LeetCode/526-beautiful-arrangement/ex_526.py
class Solution(object):
    def countArrangement(self, N):
        """
        :type N: int
        :rtype: int
        """
        nums = range(1, N + 1)
        return self.count(nums, N)
        
    def count(self, nums, n):
        if n <= 0:  return 1
        cnt = 0
        for i in range(n):
            if nums[i] % n == 0 or n % nums[i] == 0:
                nums[i], nums[n-1] = nums[n-1], nums[i]
                cnt += self.count(nums, n-1)
                nums[i], nums[n-1] = nums[n-1], nums[i]
        return cnt